models/vgg/vgg.hpp

```diff
@@ -0,0 +1,181 @@
+ *  year = {2014},
+ *  url = {https://arxiv.org/pdf/1409.1556.pdf}
+ * }
+ * @endcode
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MODELS_MODELS_VGG_VGG_HPP
+#define MODELS_MODELS_VGG_VGG_HPP
+
+#include <mlpack/core.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/layer/layer_types.hpp>
+#include <mlpack/methods/ann/init_rules/random_init.hpp>
+#include <mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp>
+
+namespace mlpack {
+namespace models {
+
+/**
+ * Definition of a VGG CNN.
+ * 
+ * NOTE: Note that output size will be 1x1xN. Here, N is number of classes.
+ * 
+ * @tparam MatType Matrix representation to accept as input and use for
+ *    computation.
+ * @tparam VGGVersion Version of VGG.
+ * @tparam IsBatchNorm Whether to apply Batch Norm layer.
+ */
+template<
+  typename MatType = arma::mat,
+  size_t VGGVersion = 11,
+  bool IsBatchNorm = false
+>
+class VGGType : public ann::MultiLayer<MatType>
+{
+ public:
+  //! Create the VGGType layer.
+  VGGType();
+
+  /**
+   * VGGType constructor intializes number of classes and weights.
+   *
+   * @param numClasses Optional number of classes to classify images into,
+   *     only to be specified if includeTop is  true.
+   * @param includeTop Must be set to true if weights are set.
+   */
+  VGGType(
+    const size_t numClasses,
+    const bool includeTop = true);
+
+  //! Copy the given VGGType.
+  VGGType(const VGGType& other);
+  //! Take ownership of the layers of the given VGGType.
+  VGGType(VGGType&& other);
+  //! Copy the given VGGType.
+  VGGType& operator=(const VGGType& other);
+  //! Take ownership of the given VGGType.
+  VGGType& operator=(VGGType&& other);
+
+  //! Virtual destructor: delete all held layers.
+  virtual ~VGGType()
+  { /* Nothing to do here. */ }
+
+  //! Create a copy of the VGGType (this is safe for polymorphic use).
+  VGGType* Clone() const { return new VGGType(*this); }
+
+  /**
+   * Get Layers of the model.
+   * 
+   * @tparam OutputLayerType The output layer type used to evaluate the network.
+   * @tparam InitializationRuleType Rule used to initialize the weight matrix.
+   */
+  template<
+    typename OutputLayerType = ann::CrossEntropyError,
+    typename InitializationRuleType = ann::RandomInitialization
+  >
+  ann::FFN<OutputLayerType, InitializationRuleType, MatType>* GetModel()
+  {
+    ann::FFN<OutputLayerType, InitializationRuleType, MatType>* vgg =
+        new ann::FFN<OutputLayerType, InitializationRuleType, MatType>();
+    vgg->Add(this);
+    return vgg;
+  }
+
+  //! Serialize the VGGType.
+  template<typename Archive>
+  void serialize(Archive& ar, const uint32_t /* version */);
+
+ private:
+  void makeModel()
```

Comment:
```suggestion
  void MakeModel()
```
Just a stylistic convention. :)  Also I would move the implementation into `vgg_impl.hpp` to keep the header file shorter.